gateway/reverse_proxy.go

```diff
@@ -789,33 +789,69 @@ const (
 	inMemNetworkType = "memu"
 )
 
-func handleInMemoryLoop(handler http.Handler, r *http.Request) (resp *http.Response, err error) {
-	// use separate provider for each request
-	provider := memconn.Provider{}
+var (
+	memconnProvidersMtx sync.Mutex
```

Comment:
Did you consider using a RWMutex? I'm asuming the providers are mostly read-only after they are created.